src/mscorlib/shared/System/SpanHelpers.Char.cs

```diff
@@ -84,79 +84,86 @@ public static unsafe int IndexOf(ref char searchSpace, char value, int length)
         {
             Debug.Assert(length >= 0);
 
-            uint uValue = value; // Use uint for comparisons to avoid unnecessary 8->32 extensions
-            IntPtr index = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
-            IntPtr nLength = (IntPtr)length;
-#if !netstandard11
-            if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
+            fixed (char* pChars = &searchSpace)
```

Comment:
Is the version with pinning faster than what used to be here before?

Reply:
Yes, slightly faster for small inputs (at least for x64). I haven't investigated on why exactly, yet.

Reply:
Also, when @eerhardt did the string.IndexOf vectorization, he didn't find any perf benefit of using IntPtr for math to avoid the 64->32->64 truncations.

Reply:
> Is the version with pinning faster than what used to be here before?

I am assuming the reason for this is the extra additions in the previous version during SequentialScan.

```C#
            while ((byte*)nLength > (byte*)0)
            {
                nLength -= 1;

                if (uValue == Unsafe.Add(ref searchSpace, index))
                    goto Found;

                index += 1;
            }
```

vs.
```C#
            while (count > 0)
            {
                if (*pCh == value)
                    goto ReturnIndex;

                count--;
                pCh++;
            }
```

Same with the first check within this loop: `while ((byte*)nLength >= (byte*)4)`

Reply:
IntPtr for math only helps in together with Unsafe.Add. It has no benefit once you start incrementing pointers.

Reply:
> IntPtr for math only helps in together with Unsafe.Add. It has no benefit once you start incrementing pointers.

Yes, I get that. I was pointing out that the IntPtr/Unsafe.Add combo (from the previous version) has extra addition instructions that the pointer version doesn't seem to have, which could be partly why the pinned version is faster.

Reply:
There should not be extra addition instruction if the JIT optimizations worked as expected. The addition should have been folded into addressing mode. Still, it will be a longer instruction at least.

Reply:
FWIW, I was always wondering whether the convoluted code that avoids pinning is worth it for larger methods. I think I am happy to see that this is proving that it is not worth it, and that it may even have negative impact. We may want to get rid of it in more places.

Avoiding pinning is definitely worth it for small/inlineable methods, but it start having diminishing and even negative returns for larger ones.

Reply:
> There should not be extra addition instruction if the JIT optimizations worked as expected. The addition should have been folded into addressing mode. Still, it will be a longer instruction at least.

I will confirm this tomorrow, and if true, maybe revert back to the original implementation.

Reply:
> @eerhardt, did you compare the performance difference between IntPtr/Unsafe.Add versus using fixed/pointers when you did the port from Span<byte>.IndexOf?

I didn't really test it deeply. One of my goals when I did that port was to keep the existing `string.IndexOf` sequential scan logic, which meant using fixed/pointers.  So it didn't really pop up on my radar.

Reply:
> There should not be extra addition instruction if the JIT optimizations worked as expected. The addition should have been folded into addressing mode. Still, it will be a longer instruction at least.

> I will confirm this tomorrow

Yes, that is exactly what's happening (no extra addition instructions).
**Using pointers:**
```text
movzx       edx,word ptr [r9] // loop start
cmp         edx,eax
je          00007FFC0BC734EA
dec         r8d
add         r9,2
test        r8d,r8d
jg          00007FFC0BC734D6
```
**Using ref and Unsafe.Add:**
```text
dec         r8	// loop start
movzx       r9d,word ptr [rcx+rdx*2]
cmp         r9d,eax
je          00007FFC0ED1349F
inc         rdx
test        r8,r8
jne         00007FFC0ED1348A
```

Reply:
It does look like the other loop (which decrements by 4) has a difference (extra lea instructions):
Red (using ref) versus Green (using pointers)
![image](https://user-images.githubusercontent.com/6527137/38055107-a438478c-328d-11e8-82b9-31cd8ec3e06d.png)

**Code snippet:**

Red
```C#
while ((byte*)nLength >= (byte*)4)
{
    nLength -= 4;

    if (uValue == Unsafe.Add(ref searchSpace, index))
        goto Found;
    if (uValue == Unsafe.Add(ref searchSpace, index + 1))
        goto Found1;
    if (uValue == Unsafe.Add(ref searchSpace, index + 2))
        goto Found2;
    if (uValue == Unsafe.Add(ref searchSpace, index + 3))
        goto Found3;

    index += 4;
}
```

Green
```C#
while (length >= 4)
{
    length -= 4;

    if (*pCh == value)
        goto Found;
    if (*(pCh + 1) == value)
        goto Found1;
    if (*(pCh + 2) == value)
        goto Found2;
    if (*(pCh + 3) == value)
        goto Found3;

    pCh += 4;
}
```

cc @AndyAyersMS

Reply:
In the red case the jit does not know it can distribute multiply over add to fit `base + 2 * (index + d)` into an addressing mode of of the form `base + 2*index + k`. 

So in the red case the add in parentheses happens up front in the `lea` instructions.

Reply:
Can you think of a way to re-write it in C# where we can make the JIT aware of that and hence avoid the `lea` instructions similar to the pointer arithmetic?

Reply:
Well, yeah, you have an example already.

But I suppose you are asking how to do this without using pointers. You'd need to express everything as accessing bytes I think.

But I suspect that extra `lea` shouldn't really hurt perf much. Do you see an impact?

Reply:
> Well, yeah, you have an example already. But I suppose you are asking how to do this without using pointers.

Right :)

> But I suspect that extra `lea` shouldn't really hurt perf much. Do you see an impact?

I observed a performance difference for small strings. Given the rest of the implementation is essentially identical (particularly for small strings, since they don't enter the vectorized code paths), I am assuming the extra `lea` are the cause of the difference.

From comment above - https://github.com/dotnet/coreclr/pull/17284#discussion_r177635478: 
![image](https://user-images.githubusercontent.com/6527137/38008695-f02e33a6-3204-11e8-9956-e25551a5cd39.png)

Interestingly though, if I rip out the vectorized code (keeping the checks), and follow the disassembly for length == 1, the ref version has fewer instructions.
![image](https://user-images.githubusercontent.com/6527137/38065342-a88ca178-32b7-11e8-9226-55d670d12aae.png)

**Red (ref):**
```C#
public static unsafe int Test1(ref char searchSpace, char value, int length)
{
    uint uValue = value; // Use uint for comparisons to avoid unnecessary 8->32 extensions
    IntPtr index = (IntPtr)0; // Use IntPtr for arithmetic to avoid unnecessary 64->32->64 truncations
    IntPtr nLength = (IntPtr)length;

    if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
    {
    }

    while ((byte*)nLength >= (byte*)4)
    {
        nLength -= 4;

        if (uValue == Unsafe.Add(ref searchSpace, index))
            goto Found;
        if (uValue == Unsafe.Add(ref searchSpace, index + 1))
            goto Found1;
        if (uValue == Unsafe.Add(ref searchSpace, index + 2))
            goto Found2;
        if (uValue == Unsafe.Add(ref searchSpace, index + 3))
            goto Found3;

        index += 4;
    }

    while ((byte*)nLength > (byte*)0)
    {
        nLength -= 1;

        if (uValue == Unsafe.Add(ref searchSpace, index))
            goto Found;

        index += 1;
    }

    if (Vector.IsHardwareAccelerated && ((int)(byte*)index < length))
    {
    }

    return -1;
    Found: // Workaround for https://github.com/dotnet/coreclr/issues/13549
    return (int)(byte*)index;
    Found1:
    return (int)(byte*)(index + 1);
    Found2:
    return (int)(byte*)(index + 2);
    Found3:
    return (int)(byte*)(index + 3);
}
```
**Green (pointer):**
```C#
fixed (char* pChars = &searchSpace)
{
    char* pCh = pChars;
    char* pEndCh = pCh + length;

    if (Vector.IsHardwareAccelerated && length >= Vector<ushort>.Count * 2)
    {
    }

    while (length >= 4)
    {
        length -= 4;

        if (*pCh == value)
            goto Found;
        if (*(pCh + 1) == value)
            goto Found1;
        if (*(pCh + 2) == value)
            goto Found2;
        if (*(pCh + 3) == value)
            goto Found3;

        pCh += 4;
    }

    while (length > 0)
    {
        length -= 1;

        if (*pCh == value)
            goto Found;

        pCh += 1;
    }

    if (Vector.IsHardwareAccelerated && pCh < pEndCh)
    {
    }

    return -1;
    Found3:
    pCh++;
    Found2:
    pCh++;
    Found1:
    pCh++;
    Found:
    return (int)(pCh - pChars);
}
```